### DISCUSSION TITLE
About setdihedral angles with RDKit

### DISCUSSION THREAD

**Discussion by TianmingQu** [rdkit/rdkit#6823] [Q&A]
Hi all, 

I am new to RDKit and I am recently trying to change all of the backbone dihedral angles in my protein and then get the position for the new_generated protein PDB file.

Here is my question, so in the current 'setdihedraldeg' function of RDKit, it can only take into four atoms of one dihedral angle, means I can only modify the dihedral angle one by one if I am correct, however, if two neighbor dihedrals share three atoms, after we change the first dihedral angle, its neighbor has already been changed, which is not what I want.

So is there anyway I can ensure that the dihedral angle added to each dihedral is exactly what I want without the influence of 'sharing neighbor atoms'? 
Like for instance, my current idea is that :

If I change the first dihedral (1-2-3-4), then I have to move the fifth atom next to this dihedral (number 5) to make sure that (2-3-4-5) is not infulenced by the change of (1-2-3-4);

And then move number 5 to change dihedral (2-3-4-5)

And the same for all the following atoms.

Is there a function, some restrain or helper function in RDKit that can help me to do this?

**Comment by jim-furness** [comment]
I'd like to second this question.

I'm struggling figure out how to set a dihedral angle `A-B-C-D` without affecting other atoms bonded to `C`, it seems as if RDKit is rotating around the `B-C` bond the behavior seems contrary to the docs at http://www.rdkit.org/docs/source/rdkit.Chem.rdMolTransforms.html#rdkit.Chem.rdMolTransforms.SetDihedralDeg which suggests only things bonded to the final atom D are moved.

e.g. if I have `[H:1][C:2][C:3]([F:4])([Cl:5])[O:6]` and I am interested in creating a slightly distorted molecule by setting e.g. the dihedrals `[H:1][C:2][C:3]([F:4]) = 20`, `[H:1][C:2][C:3]([Cl:5]) = 150`, and `[H:1][C:2][C:3]([O:6]) = 210`. What I find is that if I set these in order, then the dihedrals to `[F:4]` and `[Cl:5]` are overwritten by the change for `[O:6]`. As if the `B-C` bond is being rotated each time.

One idea I had to work around this was to break all bonds from C not included in the dihedral, then remake them after the operation. But this is resulting in a Pre-conditon Violation of `RingInfo not initialized`.

For a minimal example see,
```py
from rdkit.Chem import rdMolTransforms
from rdkit import Chem
from rdkit.Chem import AllChem

smi = 'FCCO'

mol = Chem.MolFromSmiles(smi)
mol = Chem.AddHs(mol)
AllChem.EmbedMolecule(mol)
conf = mol.GetConformer(0)

for atom in mol.GetAtoms():
    print(f'{atom.GetSymbol()}{atom.GetIdx()}')

# F-C-C-O dihedral
original = rdMolTransforms.GetDihedralDeg(conf, 0, 1, 2, 3)
print(f'F-C-C-O dihedral = {original}')

# F-C-C-H dihedral
pre = rdMolTransforms.GetDihedralDeg(conf, 0, 1, 2, 6)
print(f'F-C-C-H dihedral = {pre}')

# Change the F-C-C-O dihedral to 90 degrees
print("Changing F-C-C-O dihedral to 90.")
rdMolTransforms.SetDihedralDeg(conf, 0, 1, 2, 3, 90.0)

# re-measure the F-C-C-H dihedral
post = rdMolTransforms.GetDihedralDeg(conf, 0, 1, 2, 6)
print(f'F-C-C-H dihedral now = {post}')
```

which prints:

```
F0
C1
C2
O3
H4
H5
H6
H7
H8
F-C-C-O dihedral = 59.99993104866751
F-C-C-H dihedral = -58.003420258788076
Changing F-C-C-O dihedral to 90.
F-C-C-H dihedral now = -28.00335130745561
```

**Comment by jim-furness** [comment]
After some pointers I've found a workaround that allows one to rotatate a single dihedral in a conformation while leaving other branches from the bond unchanged.

Adapting the previous minimal example, this achieves the intended result:

```py
from rdkit.Chem import rdMolTransforms
from rdkit import Chem
from rdkit.Chem import AllChem


smi = 'FCCO'

mol = Chem.MolFromSmiles(smi)
mol = Chem.AddHs(mol)
rw_mol = Chem.RWMol(mol)
AllChem.EmbedMolecule(rw_mol)
conf = rw_mol.GetConformer(0)


for atom in rw_mol.GetAtoms():
    print(f'{atom.GetSymbol()}{atom.GetIdx()}')

# F-C-C-O dihedral
original = rdMolTransforms.GetDihedralDeg(conf, 0, 1, 2, 3)
print(f'F-C-C-O dihedral = {original}')

# F-C-C-H dihedral
pre = rdMolTransforms.GetDihedralDeg(conf, 0, 1, 2, 6)
print(f'F-C-C-H dihedral = {pre}')


def setDihedralForOneBranch(rw_mol, conf, atom_indices, angle):
    """
    Set the dihedral angle for a single branch in a molecule.
    """
    removed_bonds = []
    for bnd in rw_mol.GetAtomWithIdx(atom_indices[-2]).GetBonds():
        aids = bnd.GetBeginAtomIdx(), bnd.GetEndAtomIdx()
        if aids[0] in atom_indices and aids[1] in atom_indices:
            continue
        rw_mol.RemoveBond(*aids)
        removed_bonds.append(aids + (bnd.GetBondType(),))
    # don't need the SSSR, just to know if bonds are in rings:
    Chem.FastFindRings(rw_mol)
    rdMolTransforms.SetDihedralDeg(conf, *atom_indices, angle)
    for bnd in removed_bonds:
        rw_mol.AddBond(*bnd)

print("before")
print(Chem.MolToXYZBlock(rw_mol))

print("Changing F-C-C-O dihedral to 90.")
setDihedralForOneBranch(rw_mol, conf, [0, 1, 2, 3], 90.0)
print("after")
print(Chem.MolToXYZBlock(rw_mol))

# re-measure the F-C-C-H dihedral
post = rdMolTransforms.GetDihedralDeg(conf, 0, 1, 2, 6)
print(f'F-C-C-H dihedral now = {post}')
```

**Comment by greglandrum** [comment]
To clarify what is happening here, because it is the intended behavior.
Using this molecule;
![image](https://github.com/user-attachments/assets/bb0a1d32-8f3c-4abe-a6cc-7cedef8bbb5f)
If I change the dihedral `0-1-2-3`, the atoms which are moved are atom 3 and the atoms bound to it that aren't part of the dihedral. Here that means the moving atoms are atoms 3,4,5, and 6. You can see that here:
```
In [7]: print(Chem.MolToXYZBlock(m))
7

O      1.902101    0.706574    0.971617
C      1.899453    0.142306   -0.287258
C      0.606726   -0.630036   -0.455252
C     -0.627001    0.193374   -0.327454
C     -1.810181   -0.762243   -0.527214
F     -0.695453    1.137252   -1.332680
C     -0.822591    0.848358    0.998238


In [8]: rdMolTransforms.SetDihedralDeg(m.GetConformer(),0,1,2,3,90)

In [9]: print(Chem.MolToXYZBlock(m))
7

O      1.902101    0.706574    0.971617
C      1.899453    0.142306   -0.287258
C      0.606726   -0.630036   -0.455252
C     -0.519280    0.163645   -1.019698
C     -1.719610   -0.787267   -1.109113
F     -0.234324    0.577082   -2.305754
C     -0.954396    1.328829   -0.196446
```
I believe that this is consistent with the documentation: only atoms connected to the final atom in the dihedral are moved.

If anyone has suggestions on how we can make the docs clearer, I'm happy to hear them. :-)

**Reply by jim-furness** [reply]
I think my example was not clear because of implicit/explicit hydrogen, or maybe I am still confused 🙂.

If we take the SMILES `FCC(F)(Cl)OC`  and ignore all hydrogen:  

![example](https://github.com/user-attachments/assets/982d3392-fff6-49d4-8305-7da3fec9ac4a)

If I adjust the dihedral angle for the quad `(F0, C1, C2, O5)` from the docs I would expect only atoms directly connected to the final atom `O5` to move, in this case `C6` (along with `O5` of course). I would not expect `F3` or `Cl4` to move as these are not directly connected to the final atom `O5`.

If I run some simple code to test this:

```py
from rdkit.Chem import rdMolTransforms
from rdkit import Chem
from rdkit.Chem import AllChem

smi = 'FCC(F)(Cl)OC'

mol = Chem.MolFromSmiles(smi)
mol = Chem.AddHs(mol)
AllChem.EmbedMolecule(mol)
conf = mol.GetConformer(0)

for atom in mol.GetAtoms():
    print(f'{atom.GetSymbol()}{atom.GetIdx()}')

print(Chem.MolToXYZBlock(mol))

# F-C-C-O dihedral
original = rdMolTransforms.GetDihedralDeg(conf, 0, 1, 2, 5)
print(f'F-C-C-O dihedral = {original}')

# F-C-C-Cl dihedral
pre = rdMolTransforms.GetDihedralDeg(conf, 0, 1, 2, 4)
print(f'F-C-C-Cl dihedral = {pre}')

# Change the F-C-C-O dihedral to 90 degrees
print("Changing F-C-C-O dihedral to 90.")
rdMolTransforms.SetDihedralDeg(conf, 0, 1, 2, 5, 90.0)

# re-measure the F-C-C-Cl dihedral
post = rdMolTransforms.GetDihedralDeg(conf, 0, 1, 2, 4)
print(f'F-C-C-Cl dihedral now = {post}')

print(Chem.MolToXYZBlock(mol))
```

I get the output:

```
F0
C1
C2
F3
Cl4
O5
C6
H7
H8
H9
H10
H11
12

F     -2.568729   -0.222930    0.119309
C     -1.297042   -0.054480   -0.405181
C     -0.380951    0.287335    0.720473
F     -0.406463   -0.752131    1.625525
Cl    -1.148146    1.665324    1.560263
O      0.869629    0.696139    0.373499
C      1.636350   -0.189017   -0.346489
H     -1.412464    0.800017   -1.126398
H     -1.049261   -0.963747   -0.950257
H      2.613276    0.307088   -0.538770
H      1.892244   -1.073271    0.283296
H      1.251558   -0.500325   -1.315271

F-C-C-O dihedral = -169.99075612233457
F-C-C-Cl dihedral = -52.36977139635683
Changing F-C-C-O dihedral to 90.
F-C-C-Cl dihedral now = -152.37901527402227
12

F     -2.568729   -0.222930    0.119309
C     -1.297042   -0.054480   -0.405181
C     -0.380951    0.287335    0.720473
F     -1.044037    1.150095    1.566773
Cl     0.932400    1.255099   -0.008381
O      0.216760   -0.761923    1.347579
C     -0.610654   -1.687843    1.937325
H     -1.412464    0.800017   -1.126398
H     -1.049261   -0.963747   -0.950257
H      0.042004   -2.463659    2.395174
H     -1.144632   -1.227647    2.801677
H     -1.305949   -2.213455    1.286314


```

Showing that `Cl4` and `F3` have moved, despite not being directly connected to the final atom `O5`. Only `F0`, `C1`, and `C2` (and their associated hydrogen) are in their original positions.

Hopefully this makes my confusion clearer?

**Comment by jasondbiggs** [comment]
The docs say 

>Sets the dihedral angle in radians between atoms i, j, k, l; all atoms bonded to atom l are moved

Shouldn't they instead say something like

> all atoms bonded to atom k, excluding atom j, are moved

Using the `FCC(F)(Cl)OC` example above, if you adjust the `(F0, C1, C2, O5)` angle it would most certainly rotate every atom connected to `C2`

**Reply by jim-furness** [reply]
>   all atoms bonded to atom k, excluding atom j, are moved

This matches my understanding from experimenting and I think would be clearer. Looking at the implementation the code comments appear to agree: https://github.com/rdkit/rdkit/blob/0be09f8e7055879f714a802ca2590572f8036bf8/Code/GraphMol/MolTransforms/MolTransforms.cpp#L657

I would very much appreciate the addition of a function that cleanly implemented the operation of moving only atoms bonded to `l` excluding `k` such that an inefficient workaround like 

```py
def setDihedralForOneBranch(rw_mol, conf, atom_indices, angle):
    """ 
    Set the dihedral angle for a single branch in a molecule.
    """
    removed_bonds = []
    for bnd in rw_mol.GetAtomWithIdx(atom_indices[-2]).GetBonds():
        aids = bnd.GetBeginAtomIdx(), bnd.GetEndAtomIdx()
        if aids[0] in atom_indices and aids[1] in atom_indices:
            continue
        rw_mol.RemoveBond(*aids)
        removed_bonds.append(aids + (bnd.GetBondType(),))
    # don't need the SSSR, just to know if bonds are in rings:
    Chem.FastFindRings(rw_mol)
    rdMolTransforms.SetDihedralDeg(conf, *atom_indices, angle)
    for bnd in removed_bonds:
        rw_mol.AddBond(*bnd)
```

was not necessary.

I may try to make a PR for this myself but the best practices for how to implement it are unclear to me. 

It would seem that an additional argument flagging that the to-move atom list should be constructed only for `l` here https://github.com/rdkit/rdkit/blob/0be09f8e7055879f714a802ca2590572f8036bf8/Code/GraphMol/MolTransforms/MolTransforms.cpp#L659 is all that is needed. 

But the best way to do this such that the existing function signature is unchanged for compatibility is not clear to me. If this is a trivial implementation for someone more experienced I would welcome it.
